### PR TITLE
Teleport discover RFD

### PR DESCRIPTION
**Pull request by r0mant**

### CODE REVIEW

**Review by r0mant — COMMENTED** [review]
After talking with @klizhentas, we'd want `teleport discover` to better support UI-driven flow which means it'll likely need to run as an agent and report the progress to the cluster somehow. I'll be making updates to the RFD to reflect that.

Also need to explore the following:

- teleport discover authentication with the cluster. Exchange token for an identity like an agent?
- Post-discover options for starting Teleport: generate config file, give teleport start command, give Helm install command, give cloud-init script.

File: rfd/0073-discover.md
Comment (tigrato) [inline rfd/0073-discover.md:51]: Is it possible to drop the requirement of having an EC2 machine to run this?

If possible, this will be a huge improvement for the EKS enrollment process. 

Right now the IAM Role mapping into Kubernetes Role is done by appending extra configuration in the `configmap/aws-auth`. 
If  `teleport discover` is running in an EC2 machine, it requires that the operator must add, manually, the EC2 node role into `configmap/aws-auth`  which does not solve our issues. 

If we drop the requirement for at least the Kubernetes discovery, we can run the EKS discovery using the tokens from the user, we can grab the `AWS_PROFILE`, and it will inherit the user credentials giving Teleport the required access level to the cluster. For this solution, we should install the helm chart in the cluster. 

The `teleport discovery` is not required to start the daemon process, it can run while Teleport installs the Helm charts and exits after every cluster is ready. If the operator wants to enroll more clusters he runs the command multiple times.

Otherwise, if the operator wants to run with `--enable-discovery`, he must manually edit the `configmap/aws-auth` in order to map the EC2 IAM role into the Kubernetes role but only for the new clusters, since the old ones were already enrolled using his credentials.

What do you think?

Comment (xinding33) [reply rfd/0073-discover.md:52]: From a UX standpoint, it'd be great if we can remove the EC2 requirement but then where would we expect users to run these commands and what permissions (and accompanying security scope) would we inherit?

File: rfd/0073-discover.md
Comment (xinding33) [inline rfd/0073-discover.md:90]: I think it's worth mentioning that we want to take an opt-out approach to resource selection, meaning that by default, we should aim to discover and import all discoverable resources. This has two effects:

1. The command `teleport discover aws` with no additional arguments should seek to discover all supported resources
2. The UI should select all resource types by default and users can deselect resource types thereby opting out

File: rfd/0073-discover.md
Comment (xinding33) [inline rfd/0073-discover.md:108]: Similar to above, we should preselect all discovered resources and have users opt out instead of opt in.

File: rfd/0073-discover.md
Comment (xinding33) [inline rfd/0073-discover.md:53]: One major problem that I foresee is that users will forget to (or otherwise unable to) configure the right IAM permissions. This effectively will move the complexity in configuring resource for Teleport to complexity of configuring an EC2 instance with the right IAM permissions. Beyond documentation, what else can we do to streamline this process? Can we somehow template the EC2 instance for them with the right IAM permissions? Is there any automation we can apply here?

**Review by xinding33 — COMMENTED** [review]
Summary of my comments:

1. teleport discover (both CLI and UI) should be opt-out instead of opt-in.
2. Is there anything we can do to streamline the EC2 + IAM configuration process? I foresee this to be the biggest new blocker.

File: rfd/0073-discover.md
Comment (klizhentas) [inline rfd/0073-discover.md:45]: The phase 2 will support RDS as well?

Comment (r0mant) [reply rfd/0073-discover.md:45]: Yes, the idea is to split implementation in phases and implement the discover flow for EC2 instances first, then RDS and other databases. I'll make implementation phases more explicit in the RFD.

File: rfd/0073-discover.md
Comment (klizhentas) [reply rfd/0073-discover.md:90]: Agree with @xinding33 here. Let's do opt-out. We can find and discover all resources and let users to opt out of resources they don't need by using labels.

File: rfd/0073-discover.md
Comment (klizhentas) [inline rfd/0073-discover.md:153]: Can this command fail gracefully, for example

[1] Checking your AWS permissions
[2] This user is missing permissions, .... apply this policy for user to proceed

File: rfd/0073-discover.md
Comment (klizhentas) [inline rfd/0073-discover.md:247]: @r0mant you mentioned that you can let users to keep the service running. Can teleport discover generate systemd file, apt-get? Basically can we show users how to keep this stick without going to the docs? Something like that:

```
$ teleport discover isntall

[1] Generated `/etc/teleport.yaml`
[2] Generated `/etc/systemd.d/teleport.service

Run `sudo systemctl teleport start` to keep Teleport agent running. For HA or K8s setup follow this guide:

....

```

File: rfd/0073-discover.md
Comment (xinding33) [reply rfd/0073-discover.md:53]: This would be an awesome improvement!

File: rfd/0073-discover.md
Comment (tigrato) [reply rfd/0073-discover.md]: My initial idea was to use `teleport discover aws` to simplify the current infrastructure onboarding with a minimal amount of actions even for cases where the infrastructure is static and continuous discovery is not something that fits the client.

When running the CLI locally, it  could have two modes:

- Automatic: After the administrator decided which resources to onboard, Teleport inherits the user credentials and does everything automatically. This includes the setup of required IAM policies for Teleport to fully work. Although the process is automatic the actions are shown to the user so he can review them. 
```
[1] Connecting to the AWS environment...
[2] Checking your user permissions....
...
[n] Onboarding database...
The teleport cluster lacks the following permissions to onboard the database ...
`json policy`

Do you accept the automatic creation of the policy?
(Y) Yes
(M) Manual
(N) No

...
```
- Manual:  we show the action required as Sasha mentioned:
```go
$ aws iam assign policy < `json policy`

or use this IAM JSON blob...
```

The same applies to other resources like EKS clusters (edit of `configmap/aws-auth`), nodes... 

Using Automatic mode has several security concerns, but if we make it in a way where users control the process  and they run it locally for short periods of time it would simplify a lot the onboarding process.

File: rfd/0073-discover.md
Comment (klizhentas) [inline rfd/0073-discover.md:124]: nit: let's please change Automatic to something else - when I saw it after a while, I got confused - is AWS not automatic? What's the difference?

File: rfd/0073-discover.md
Comment (klizhentas) [inline rfd/0073-discover.md:141]: Let's invite user to install Teleport connect here. It offers a much better US for getting started, especially for databases

**Review by klizhentas — APPROVED** [review]

File: rfd/0073-discover.md
Comment (klizhentas) [inline rfd/0073-discover.md:188]: I think this flow looks good, but needs a bit more detail on specifics of each stage.

**Review by klizhentas — APPROVED** [review]

File: rfd/0073-discover.md
Comment (ibeckermayer) [inline rfd/0073-discover.md:36]:
```suggestion
Improvements proposed in this RFD aim to take advantage of the existing auto
```

**Review by ibeckermayer — COMMENTED** [review]
I left a couple of comments related to places where we might educate users on what is going on with Teleport under-the-hood. The overarching ethos of my comments is based on the[ Unix philosophy](https://en.wikipedia.org/wiki/Unix_philosophy). Based on that article, the philosophy remains controversial and isn't well defined, but some of the principles that I think are relevant (with qualification) are
- Write transparent programs
- Write flexible and open programs
- Avoid captive user interfaces

I'm not opposed to this "captive user interface" outright (and agree its a good feature to add), what I'm arguing is that any GUI configuration should be treated as a supplement to our text/cli configuration, and we should make a concerted effort to educate the user on how the two relate to one another. 

As a point of comparison for what I'm saying we should try to avoid, see the tedious, GUI-constrained buffoonery Microsoft forces us to contend with in [setting up Desktop Access](https://goteleport.com/docs/desktop-access/getting-started/#step-27-prevent-the-service-account-from-performing-interactive-logins). Obviously those GUI steps are all configuring some registers/files/databases under-the-hood, but because Microsoft doesn't care to expose that information to the user, we're forced to walk our users to their bogus, un-scriptable tooling.

Granted that even if my point is generally agreed upon, there's still a judgment call to be made about how deep into the weeds we should go on any given topic in the "Day 1" wizard. My particular suggestions are just my speculations on what would help a Day 1 user gain a deeper understanding, but of course they also make the wizard take slightly longer to get through.

**Comment by zmb3** [comment]
Hey team - what's left to do before we can merge this? We've done quite a bit of work on Teleport Discover so far, so it would be great if we could ensure this is up to date and get it merged.

**Review by russjones — CHANGES_REQUESTED** [review]
Block merge.

**Review by russjones — APPROVED** [review]
lgtm

**Pull request merged**
